### PR TITLE
Use KMP-Crypto instead of half-baked internal crypto helpers

### PR DESCRIPTION
**Pull request by JesusMcCloud**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,6 @@
 [submodule "conventions-vclib/gradle-conventions-plugin"]
 	path = conventions-vclib/gradle-conventions-plugin
-	url = https://github.com/a-sit-plus/gradle-conventions-plugin.git
+	url = https://github.com/a-sit-plus/gradle-conventions-plugin.git	
+[submodule "kmp-crypto"]
+	path = kmp-crypto
+	url = git@github.com:a-sit-plus/kmp-crypto.git
```
Comment (nodh) [inline .gitmodules:6]: Hier wär noch die `https://` URL zu bevorzugen, für clones aus CI Runnern

File: gradle.properties [hunk 2]
```diff
@@ -18,6 +18,6 @@ kotlin.experimental.tryK2=false
 # workaround dokka bug (need to wait for next snapshot build)
 org.jetbrains.dokka.classpath.excludePlatformDependencyFiles=true
 
-artifactVersion = 3.4.0-SNAPSHOT
-
+artifactVersion = 3.0.2-SNAPSHOT
```
Comment (nodh) [inline gradle.properties:21]: Das passt nicht

File: vclib-aries/build.gradle.kts [hunk 3]
```diff
@@ -13,26 +13,31 @@ val artifactVersion: String by extra
 group = "at.asitplus.wallet"
 version = artifactVersion
 
+val iosDisabled: String? by extra
```
Comment (nodh) [inline vclib-aries/build.gradle.kts:16]: Why?

File: vclib-aries/src/commonTest/kotlin/at/asitplus/wallet/lib/aries/DummyCredentialDataProvider.kt [hunk 4]
```diff
@@ -30,7 +31,8 @@ class DummyCredentialDataProvider(
         if (credentialScheme != ConstantIndex.AtomicAttribute2023) {
             return KmmResult.failure(UnsupportedOperationException("no data"))
         }
-        val subjectId = subjectPublicKey.toJsonWebKey().identifier
+        val subjectId = subjectPublicKey.toJsonWebKey()
+            .fold(onSuccess = { it.keyId!! }, onFailure = { return KmmResult.failure(it) })
```
Comment (nodh) [inline vclib-aries/src/commonTest/kotlin/at/asitplus/wallet/lib/aries/DummyCredentialDataProvider.kt:35]: What if `keyId` is null?

File: vclib-openid/build.gradle.kts [hunk 5]
```diff
@@ -13,25 +13,34 @@ val artifactVersion: String by extra
 group = "at.asitplus.wallet"
 version = artifactVersion
 
+
+val iosDisabled: String? by extra
```
Comment (nodh) [inline vclib-openid/build.gradle.kts:17]: Why?

File: vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt [hunk 6]
```diff
@@ -220,16 +221,16 @@ class OidcSiopWallet(
         val now = clock.now()
         // we'll assume jwk-thumbprint
         val idToken = IdToken(
-            issuer = agentPublicKey.toJsonWebKey().jwkThumbprint,
-            subject = agentPublicKey.toJsonWebKey().jwkThumbprint,
-            subjectJwk = agentPublicKey.toJsonWebKey(),
+            issuer = agentPublicKey.toJsonWebKey().getOrNull()!!.jwkThumbprint,
```
Comment (nodh) [inline vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt:224]: what if this fails? combine it into one statement?

File: vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidvci/WalletService.kt [hunk 7]
```diff
@@ -101,7 +102,7 @@ class WalletService(
                 ).serialize().encodeToByteArray(),
                 addKeyId = true,
                 addJsonWebKey = true
-            )!!
+            )?.serialize()!!
```
Comment (nodh) [inline vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidvci/WalletService.kt:105]: What if this fails? Why first the null check then forced unwrapping?

File: vclib/build.gradle.kts [hunk 8]
```diff
@@ -14,19 +14,27 @@ val artifactVersion: String by extra
 group = "at.asitplus.wallet"
 version = artifactVersion
 
+
+val iosDisabled: String? by extra
```
Comment (nodh) [inline vclib/build.gradle.kts:18]: Why?

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/cbor/CoseService.kt [hunk 9]
```diff
@@ -93,15 +98,19 @@ class DefaultVerifierCoseService(
             payload = coseSigned.payload,
         ).serialize()
 
-        val algorithm = coseSigned.protectedHeader.value.algorithm
-            ?: return KmmResult.failure(IllegalArgumentException("Algorithm not specified"))
-        val publicKey = signer.toCryptoPublicKey()
-            ?: return KmmResult.failure<Boolean>(IllegalArgumentException("Signer not convertible"))
+        val algorithm = coseSigned.protectedHeader.value.algorithm ?: return KmmResult.failure(
+            IllegalArgumentException(
```
Comment (nodh) [inline vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/cbor/CoseService.kt:102]: please reformat

File: vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/cbor/CoseSerializationTest.kt [hunk 10]
```diff
@@ -11,12 +17,14 @@ import kotlinx.serialization.cbor.ByteStringWrapper
 
 class CoseSerializationTest : FreeSpec({
 
+    Napier.base(DebugAntilog())
```
Comment (nodh) [inline vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/cbor/CoseSerializationTest.kt:20]: Really needed here?

File: vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/CborSerializationTest.kt [hunk 11]
```diff
@@ -24,6 +26,10 @@ import kotlinx.serialization.encodeToString
 
 class CborSerializationTest : FreeSpec({
 
+    beforeSpec {
+        Napier.base(DebugAntilog())
```
Comment (nodh) [inline vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/CborSerializationTest.kt:30]: Really needed here?

File: vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/IsoMdocTest.kt [hunk 12]
```diff
@@ -28,6 +31,10 @@ import kotlin.random.Random
 
 class IsoMdocTest : FreeSpec({
 
+    beforeSpec {
+        Napier.base(DebugAntilog())
```
Comment (nodh) [inline vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/IsoMdocTest.kt:35]: Really needed here?

File: vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/JsonSerializationTest.kt [hunk 13]
```diff
@@ -17,6 +19,10 @@ import kotlinx.datetime.LocalDate
 
 class JsonSerializationTest : FreeSpec({
 
+    beforeSpec {
+        Napier.base(DebugAntilog())
```
Comment (nodh) [inline vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/JsonSerializationTest.kt:23]: Really needed here?

File: vclib/src/jvmMain/kotlin/at/asitplus/wallet/lib/agent/DefaultCryptoService.kt [hunk 14]
```diff
@@ -49,83 +37,89 @@ import kotlin.time.Duration.Companion.days
...
+
+    override suspend fun sign(input: ByteArray): KmmResult<CryptoSignature> = runCatching {
+        val sig = Signature.getInstance(algorithm.jcaName).apply {
+            this@DefaultCryptoService.algorithm.jcaParams?.let { setParameter(it) }
+            initSign(privateKey)
+            update(input)
+        }.sign()
+        //In Java EC signatures are returned as DER-encoded, RSA signatures however are raw bytearrays
+        if (algorithm.isEc)
+            CryptoSignature.decodeFromDer(sig)
```
Comment (nodh) [inline vclib/src/jvmMain/kotlin/at/asitplus/wallet/lib/agent/DefaultCryptoService.kt:115]: Would it be a good idea to define a function like `CryptoSignature.parseFromJca(bytes: ByteArray)` to handle that case?

**Review by nodh — CHANGES_REQUESTED** [review]

File: vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/CborSerializationTest.kt (hunk 11 above)
Comment (JesusMcCloud) [reply vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/CborSerializationTest.kt:30]: should be moved to global test configuration, yes!

File: vclib/src/jvmMain/kotlin/at/asitplus/wallet/lib/agent/DefaultCryptoService.kt (hunk 14 above)
Comment (JesusMcCloud) [reply vclib/src/jvmMain/kotlin/at/asitplus/wallet/lib/agent/DefaultCryptoService.kt:115]: probably, but this should then go into a feature request for KMP-Crypto

File: vclib-aries/build.gradle.kts (hunk 3 above)
Comment (JesusMcCloud) [reply vclib-aries/build.gradle.kts:16]: to work around an IDEA bug. it is now documented in DEVELOPMENT.md

File: vclib-openid/build.gradle.kts (hunk 5 above)
Comment (JesusMcCloud) [reply vclib-openid/build.gradle.kts:17]: to work around an IDEA bug. it is now documented in DEVELOPMENT.md

File: gradle.properties (hunk 2 above)
Comment (JesusMcCloud) [reply gradle.properties:21]: fixed

File: vclib/src/jvmMain/kotlin/at/asitplus/wallet/lib/agent/DefaultCryptoService.kt (hunk 14 above)
Comment (n0900) [reply vclib/src/jvmMain/kotlin/at/asitplus/wallet/lib/agent/DefaultCryptoService.kt:115]: can now be found at `feature/parseJcaSignature` in KMP-Crypto

File: vclib-aries/src/commonTest/kotlin/at/asitplus/wallet/lib/aries/DummyCredentialDataProvider.kt (hunk 4 above)
Comment (n0900) [reply vclib-aries/src/commonTest/kotlin/at/asitplus/wallet/lib/aries/DummyCredentialDataProvider.kt:35]: This should still be identifier. Thanks

Comment (JesusMcCloud) [reply vclib-aries/src/commonTest/kotlin/at/asitplus/wallet/lib/aries/DummyCredentialDataProvider.kt:35]: tricky one. this issue was always there, but hidden away

File: vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/CborSerializationTest.kt (hunk 11 above)
Comment (JesusMcCloud) [reply vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/iso/CborSerializationTest.kt:30]: if you only execute a single test, napier could be uninitalized.

**Review by nodh — APPROVED** [review]

**Pull request merged**